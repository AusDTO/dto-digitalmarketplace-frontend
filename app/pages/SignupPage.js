import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withRouter, Switch, Route } from 'react-router-dom'
import { required, validEmail, governmentEmail } from '../components/validators'

import BaseForm from '../components/shared/form/BaseForm'
import Textfield from '../components/shared/form/Textfield'
import formProps from '../components/shared/form/formPropsSelector'

import { handleSignupSubmit } from '../actions/memberActions'

import { SignupForm } from '../components/Signup/SignupForm'
import UserOnboardingContainer from '../components/Onboarding/OnboardingContainer'
import CreateUserPage from './CreateUserPage'
import NotFound from '../components/shared/NotFound'

class SignupPage extends BaseForm {
  static propTypes = {
    action: PropTypes.string,
    csrf_token: PropTypes.string,
    form: PropTypes.object
  }

  constructor(props) {
    super(props)
    this.state = {
      emailValidators: {
        required,
        validEmail
      },
      emailErrorMessages: {
        required: 'Your email is required',
        validEmail: 'A validly formatted email is required.',
        governmentEmail: ' Email should have a government domain.'
      },
      isBuyer: this.props.signupForm.user_type === 'buyer',
      submitClicked: null
    }
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.signupForm.user_type !== nextProps.signupForm.user_type) {
      if (nextProps.signupForm.user_type === 'seller') {
        this.setState({
          emailValidators: {
            required,
            validEmail
          },
          emailErrorMessages: {
            required: 'Your email is required',
            validEmail: 'A validly formatted email is required.'
          },
          isBuyer: false
        })
      }

      if (nextProps.signupForm.user_type === 'buyer') {
        this.setState({
          emailValidators: {
            required,
            validEmail,
            governmentEmail
          },
          emailErrorMessages: {
            required: 'Your email is required',
            validEmail: 'A validly formatted email is required.',
            governmentEmail: ' Email should have a government domain.'
          },
          isBuyer: true
        })
      }
    }
  }

  handleSubmit(model) {
    this.props.handleSignupSubmit(model)
  }

  onSubmitFailed() {
    window.scrollTo(0, 0)
  }

  onSubmitClicked = () => {
    this.setState({
      submitClicked: new Date().valueOf()
    })
  }

  render() {
    const { csrf_token, signupForm, user, model, form, buyer_url, seller_url, signupSuccess } = this.props
    let { isBuyer, emailValidators, emailErrorMessages } = this.state
    return (
      <SignupForm
        csrf_token={csrf_token}
        signupForm={signupForm}
        user={user}
        model={model}
        form={form}
        buyer_url={buyer_url}
        seller_url={seller_url}
        signupSuccess={signupSuccess}
        submitClicked={this.onSubmitClicked}
        handleSubmit={this.handleSubmit}
        submitFailed={this.onSubmitFailed}
        isBuyer={isBuyer}
        emailValidators={emailValidators}
        emailErrorMessages={emailErrorMessages}
      />
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    handleSignupSubmit: model => dispatch(handleSignupSubmit(model))
  }
}

const mapStateToProps = state => {
  return {
    ...formProps(state, 'signupForm'),
    signupSuccess: state.user.signupSuccess
  }
}

export { Textfield, mapStateToProps, SignupForm as Form }

const SignupPageContainer = connect(mapStateToProps, mapDispatchToProps)(SignupPage)

const Routes = ({ match }) =>
  <Switch>
    <Route exact path={match.url} component={SignupPageContainer} />
    <Route path={`${match.url}/createuser/:token`} component={CreateUserPage} />
    <Route path={`${match.url}/success`} component={UserOnboardingContainer} />
    <Route component={NotFound} />
  </Switch>

export default withRouter(Routes)
