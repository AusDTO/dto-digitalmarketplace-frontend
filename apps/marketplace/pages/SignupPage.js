/* eslint-disable camelcase */
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter, Switch, Route, Redirect } from 'react-router-dom'

import BaseForm from 'shared/form/BaseForm'
import formProps from 'shared/form/formPropsSelector'

import { required, validEmail, governmentEmail } from '../components/validators'
import SignupForm from '../components/SignupForm/SignupForm'
import UserOnboardingContainer from '../components/Onboarding/OnboardingContainer'
import { handleSignupSubmit } from '../actions/memberActions'
import { rootPath } from '../routes'

export class SignupPageComponent extends BaseForm {
  constructor(props) {
    super(props)
    this.state = {
      emailValidators: {
        required,
        validEmail
      },
      emailErrorMessages: {
        required: 'Your email is required',
        validEmail: 'A validly formatted email is required.',
        governmentEmail: ' Email should have a government domain.'
      },
      isBuyer: this.props.signupForm.user_type === 'buyer',
      submitClicked: null
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.signupForm.user_type !== nextProps.signupForm.user_type) {
      if (nextProps.signupForm.user_type === 'seller') {
        this.setState({
          emailValidators: {
            required,
            validEmail
          },
          emailErrorMessages: {
            required: 'Your email is required',
            validEmail: 'A validly formatted email is required.'
          },
          isBuyer: false
        })
      }

      if (nextProps.signupForm.user_type === 'buyer') {
        this.setState({
          emailValidators: {
            required,
            validEmail,
            governmentEmail
          },
          emailErrorMessages: {
            required: 'Your email is required',
            validEmail: 'A validly formatted email is required.',
            governmentEmail: ' Email should have a government domain.'
          },
          isBuyer: true
        })
      }
    }
  }

  handleSubmit = model => {
    this.props.handleSignupSubmit(model)
  }

  onSubmitFailed = () => {
    window.scrollTo(0, 0)
  }

  onSubmitClicked = () => {
    this.setState({
      submitClicked: new Date().valueOf()
    })
  }

  render() {
    const { match } = this.props

    // TODO: will move UserOnboardingContainer out of here when CreateUser is refactored
    return (
      <div id="signup-page">
        <Switch>
          <Route
            exact
            path={match.url}
            render={() =>
              <SignupForm
                {...this.props}
                {...this.state}
                onSubmitClicked={this.onSubmitClicked}
                onSubmitFailed={this.onSubmitFailed}
                handleSubmit={this.handleSubmit}
                submitClicked={false}
              />}
          />
          <Route path={`${match.url}/success`} component={UserOnboardingContainer} />
          <Redirect to={{ pathname: `${rootPath}/signup` }} />
        </Switch>
      </div>
    )
  }
}

SignupPageComponent.propTypes = {
  signupSuccess: PropTypes.bool,
  currentlySending: PropTypes.bool
}

SignupPageComponent.defaultProps = {
  signupSuccess: null,
  currentlySending: null
}

const mapStateToProps = state => ({
  ...formProps(state, 'signupForm'),
  signupSuccess: state.user.signupSuccess,
  currentlySending: state.app.currentlySending
})

const mapDispatchToProps = dispatch => ({
  handleSignupSubmit: model => dispatch(handleSignupSubmit(model))
})

const SignupPage = withRouter(connect(mapStateToProps, mapDispatchToProps)(SignupPageComponent))

export default SignupPage
