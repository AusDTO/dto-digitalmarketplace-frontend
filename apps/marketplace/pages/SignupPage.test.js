import React from 'react'
import expect from 'expect'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import { Provider } from 'react-redux'

import configureStore from '../store'
import SignupPage from './SignupPage'

jest.mock('shared/Icon/_getIcons')

describe('Test suite for SignupPage component', () => {
  const state = {}

  const store = configureStore()
  test('SignupPage renders without errors', () => {
    const renderer = new ReactShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <SignupPage {...state} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })
})
