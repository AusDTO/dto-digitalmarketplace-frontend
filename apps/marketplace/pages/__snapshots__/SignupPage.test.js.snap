// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for SignupPage component SignupPage renders without errors 1`] = `<withRouter(Connect(SignupPageComponent)) />`;
